clang/include/clang/Basic/Attr.td

```diff
@@ -1226,6 +1226,17 @@ def SYCLKernel : InheritableAttr {
   let Documentation = [SYCLKernelDocs];
 }
 
+def SYCLSpecialClass: InheritableAttr {
+  let Spellings = [Clang<"sycl_special_class">];
+  let Subjects = SubjectList<[CXXRecord]>;
+  let LangOpts = [SYCLIsDevice, SYCLIsHost];
+  let Documentation = [Undocumented];
```

Comment:
Is this an internal-only attribute (e.g., users should never write this attribute in their own code)?

Reply:
That's correct.

Reply:
Okay, thanks for verifying! Comments about that may be helpful so we don't think this is an accidentally undocumented attribute.

Reply:
> If this is meant to be upstreamed, I believe they will want to see documentation.

It depends. If this is expected to be an attribute users write, upstream will definitely want documentation. If this isn't expected to be written by users, it's questionable how much upstream will want the attribute at all. (This attribute is basically needed to work around what could be design issues in SYCL, so if I had to guess, I'd guess upstream would rather see the potential design issues addressed that necessitate the attribute in the first place.)

> But regardless of the status, documentation is always better and doesn't cost much to at least describe what this attribute is meant for and what is expected for its argument. This helps newcomers to contribute.

There's no distinction in upstream between "this is for internal use" and "this is for typical use", so when you document the attribute, users are likely to expect that it's an attribute for them to use. If we're adding docs, we should make it clear that this is an internal implementation detail. (I suspect we're not consistent about this with SYCL attributes.)

Reply:
> it's questionable how much upstream will want the attribute at all

The first sycl related attribute fell under this category, yet documentation was requested:
 https://clang.llvm.org/docs/AttributeReference.html#sycl-kernel .

>  If we're adding docs, we should make it clear that this is an internal implementation detail. (I suspect we're not consistent about this with SYCL attributes.)

Definitely !

> This attribute is basically needed to work around what could be design issues in SYCL, so if I had to guess, I'd guess upstream would rather see the potential design issues addressed that necessitate the attribute in the first place.

I know what it is needed for, I suggested its addition (ComputeCpp uses a similar approach) :)
The need arises because of a different way of expressing what constitues a kernel argument in SYCL and in a SPIR-like kernel. It is a pain as an implementer but I wouldn't call that a design issue, it is just a translation problem.

Don't miss-quote me here though, there are design issues in that area (images or lambda for instance), but expressing things in a differently doesn't constitue one in itself.